error_handler.py

```diff
@@ -0,0 +1,114 @@
+import datetime
+import io
+import logging
+import traceback
+import urllib
+
+import discord
+from discord.ext import commands
+
+logger = logging.getLogger(__name__)
+
+
+def get_full_class_name(obj):
+    module = obj.__class__.__module__
+    if module is None or module == str.__class__.__module__:
+        return obj.__class__.__name__
+    return module + '.' + obj.__class__.__name__
+
+class NonBugError(Exception):
+    """When this is raised instead of a normal Exception, on_command_error() will not attach a traceback or github
+    link. """
+    pass
+
+# completely overkill error handler shamelessly ripped from MediaForge, a bot i wrote
+class ErrorHandlerCog(commands.Cog):
+    def __init__(self, bot):
+        self.bot: commands.Bot = bot
+        self.antispambucket = {}
+
+    @commands.Cog.listener()
+    async def on_command_error(self, ctx: commands.Context, commanderror: commands.CommandError):
+        async def dmauthor(*args, **kwargs):
+            try:
+                return await ctx.reply(*args, **kwargs)
+            except discord.Forbidden:
+                logger.info(f"Reply to {ctx.message.id} and dm to {ctx.author.id} failed. Aborting.")
+
+        async def reply(msg, file=None, embed=None):
+            if ctx.interaction:
+                if ctx.interaction.response.is_done():
+                    return await ctx.interaction.edit_original_response(content=msg,
+                                                                        attachments=[file] if file else None,
+                                                                        embed=embed)
+                else:
+                    return await ctx.reply(msg, file=file, embed=embed)
+            else:
+                try:
+                    if ctx.guild and not ctx.channel.permissions_for(ctx.me).send_messages:
+                        logger.debug(f"No permissions to reply to {ctx.message.id}, trying to DM author.")
+                        return await dmauthor(msg, file=file, embed=embed)
+                    return await ctx.reply(msg, file=file, embed=embed)
+                except discord.Forbidden:
+                    logger.debug(f"Forbidden to reply to {ctx.message.id}, trying to DM author")
+                    return await dmauthor(msg, file=file, embed=embed)
+
+        async def logandreply(message):
+            if ctx.guild:
+                ch = f"channel #{ctx.channel.name} ({ctx.channel.id}) in server {ctx.guild} ({ctx.guild.id})"
+            else:
+                ch = "DMs"
+            logger.info(f"Command '{ctx.message.content}' by "
+                        f"@{ctx.message.author.name}#{ctx.message.author.discriminator} ({ctx.message.author.id}) "
+                        f"in {ch} failed due to {message}.")
```

Comment:
Can we clean this up? I think it'd benefit from not using f-strings. Plus, discriminators don't exist anymore.

Reply:
...looking at how this is used, should it be using `logger.exception()`?

Reply:
> discriminators don't exist anymore.

this is (unfortunately?) false, bots still have them, and so do old unmigrated accounts, code still needs to account for it

Reply:
however in this case the default user repr accounts for it so we can just pass ctx.message.author to fmt

Reply:
> Can we clean this up? I think it'd benefit from not using f-strings. Plus, discriminators don't exist anymore.

for me, f-strings are way easier to read, especially when there's many parameters. This code was directly lifted from another bot of mine that I haven't updated in a while. 
Also I believe i used logger.info because a full logger.exception traceback is not necessary for non-errors like invalid commands